test/funder.js

```diff
@@ -0,0 +1,328 @@
+ */
+function getTransactionReceiptMined(txHash, interval) {
+  const self = this;
+  const transactionReceiptAsync = (resolve, reject) => {
+    self.getTransactionReceipt(txHash, (error, receipt) => {
+      if (error) {
+        reject(error);
+      } else if (receipt == null) {
+        setTimeout(
+          () => transactionReceiptAsync(resolve, reject),
+          interval || 500,
+        );
+      } else {
+        resolve(receipt);
+      }
+    });
+  };
+
+  if (Array.isArray(txHash) && txHash.length > 0) {
+    return Promise.all(txHash.map(
+      oneTxHash => self.getTransactionReceiptMined(oneTxHash, interval),
+    ));
+  }
+  if (typeof txHash === 'string') {
+    return new Promise(transactionReceiptAsync);
+  }
+  throw new Error(`Invalid Type: ${txHash}`);
+}
+
+/**
+ * This method captures detail about funding transaction from fauce like faucet
+ * address, eip20 token address etc. which is used to return remaining fund
+ * to faucet.
+ * @param {Array<string>} originFaucetTXHashes Transaction hashes of funding
+ *                                             transaction by mosaic faucet
+ *                                             on the origin chain.
+ * @param {Array<string>} auxiliaryFaucetTXHashes Transaction hashes of funding
+ *                                                transaction by mosaic faucet
+ *                                                on the auxiliary chain.
+ * @param {Array<string>} ropstenFaucetTXHashes Transaction hashes of funding
+ *                                              transaction by ropsten faucet
+ *                                              on the origin chain.
+ * @param {Web3} originWeb3 Web3 instance pointing to the origin chain.
+ * @param auxiliaryWeb3 Web3 instance pointing to the auxiliary chain.
+ * @return {Promise<RefundTransactionDetails>} Details of refund transaction
+ *                                            to mosaic and ropsten faucet.
+ */
+const faucetTransactionDetails = async (
+  originFaucetTXHashes,
+  auxiliaryFaucetTXHashes,
+  ropstenFaucetTXHashes,
+  originWeb3,
+  auxiliaryWeb3,
+) => {
+  // Get faucet address from one transaction.
+  const originTransaction = originWeb3.eth.getTransaction(originFaucetTXHashes[0]);
+  const auxiliaryTransaction = auxiliaryWeb3.eth.getTransaction(auxiliaryFaucetTXHashes[0]);
+  const ropstenTransaction = originWeb3.eth.getTransaction(ropstenFaucetTXHashes[0]);
+
+  return Promise.all(
+    [
+      originTransaction,
+      auxiliaryTransaction,
+      ropstenTransaction,
+    ],
+  ).then(transactions => ({
+    originTransactions: {
+      type: 'ERC20',
+      faucetAddress: transactions[0].from,
+      tokenAddress: transactions[0].to,
+
+    },
+    auxiliaryTransactions: {
+      type: 'BaseToken',
+      faucetAddress: transactions[1].from,
+    },
+    ropstenTransactions: {
+      type: 'BaseToken',
+      faucetAddress: ROPSTEN_REFUND_ADDRESS,
```

Comment:
Why is this a hardcoded constant? Can it not be taken from the transaction?

Reply:
For ropsten faucet, the transaction is made to a contract with the beneficiary address as an argument. So the contract holds the funds.  
 @benjaminbollen  shared a refund address with me where funds should be returned.